src/mode.test.ts

```diff
@@ -0,0 +1,282 @@
+import CodeMirror from 'codemirror';
+import 'codemirror/mode/meta';
+import 'codemirror/addon/mode/simple';
+import { defineRobotMode, MODE_NAME } from './mode';
+
+// https://github.com/jsdom/jsdom/issues/3002
+document.createRange = () => {
+  const range = new Range();
+
+  range.getBoundingClientRect = jest.fn();
+
+  range.getClientRects = () => {
+    return {
+      item: () => null,
+      length: 0,
+      [Symbol.iterator]: jest.fn()
+    };
+  };
+
+  return range;
+}
+
+const tokenTypeMap: { [key: string]: string } = {
+  'cm-header': 'header',
+  'cm-atom': 'atom',
+  'cm-keyword': 'keyword',
+  'cm-attribute': 'attr',
+  'cm-operator': 'operator',
+  'cm-variable-2': 'var2',
+  'cm-string': 'string',
+}
+
+const mapTokenTypes = (value: string): string => value?.split(' ').map((x) => {
+  const tokenType = tokenTypeMap[x];
+  return tokenType ? `<${tokenType}>` : x;
+}).join(' ');
+
+const findLines = (
+  editor: CodeMirror.EditorFromTextArea,
+  body: HTMLElement,
+): string[] => {
+  const lineCount = editor.lineCount();
+  const lines: string[] = [];
+  for (let line = 0; line < lineCount; line++) {
+    editor.setCursor({line: line, ch: 0});
+    const nodes = body.querySelector('.CodeMirror-scroll .CodeMirror-line span').childNodes;
+    const spans: any[] = [];
+    nodes.forEach((x) => spans.push(x));
+    const allTokens = spans.map((span) => ({
+      types: mapTokenTypes(span.attributes?.class?.value),
+      text: span.innerHTML,
+    }));
+    const tokens = allTokens
+      .filter((x) => x.text !== undefined && x.text !== '&nbsp;')
+      .map((token) => `${token.types}: ${token.text}`);
+    lines.push([`#${line}`, ...tokens].join(' '));
+  }
+  return lines;
+}
+
+defineRobotMode({ CodeMirror } as any);
+document.body.innerHTML = `<textarea id='code' />`
+const codeElement = document.getElementById('code') as HTMLTextAreaElement;
+const editor = CodeMirror.fromTextArea(codeElement, { mode: MODE_NAME });
+
+const parse = (value: string): string[] => {
+  const input = value.split('\n').filter((x) => !!x);
+  editor.setValue([
+    '*** Tasks ***',
+    'Task ATest',
+    ...input.map((x) => `    ${x}`),
+  ].join('\n'));
+  const output = findLines(editor, document.body);
+  return output.slice(2);
+}
+
+test('Tasks', () => {
```

Comment:
Awesome stuff. Given these all basically compare two static strings, would it be possible to move the literal text into standalone files, e.g. `fixtures/00-tasks.robot` and `fixtures/00-tasks.tokens`? I also don't imagine we'd need to ship these in the released tarball.

Reply:
I like the idea extracting fixtures into separate files. Although I'd put input and expected output in the same file. Then it would be easy to compare them as they sit next to each other. As there are not too many tests yet, I'd keep it as it is for the time being. Would you agree?